### ISSUE TITLE
want to loadimage in asyn 

### ISSUE THREAD

**Issue by sunforcode** [SDWebImage/SDWebImage#2539]
### New Issue Checklist

* [ ] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [ ] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [ ] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | e.g. ios / macos / tvos / watchos
 Platform Version        | e.g. 11.0 / 10.13.0 / 11.0 / 4.0
 SDWebImage Version      | e.g. 4.2.0 / 4.1.0
 Integration Method      | e.g. carthage / cocoapods / manually
 Xcode Version           | e.g. Xcode 9 / Xcode 8
 Repro rate              | e.g. all the time (100%) / sometimes x% / only once
 Repro with our demo prj | e.g. does it happen with our demo project?
 Demo project link       | e.g. link to a demo project that highlights the issue

### Issue Description and Steps

Please fill in the detailed description of the issue (full output of any stack trace, compiler error, ...) and the steps to reproduce the issue.

I want to download an image , i use   
```
 [[SDWebImageManager sharedManager]loadImageWithURL:[NSURL URLWithString:adModel.imageUrl] options:SDWebImageRefreshCached progress:nil completed:^(UIImage * _Nullable image, NSData * _Nullable data, NSError * _Nullable error, SDImageCacheType cacheType, BOOL finished, NSURL * _Nullable imageURL) {
            dispatch_semaphore_signal(self.semaphore);`  
        }];
 dispatch_semaphore_wait(self.semaphore, DISPATCH_TIME_FOREVER);,
````

it cause death lock ,i want to konw how to download image and not to go back mainThread auto

**Comment by dreampiggy** [comment]
Use Dispatch Group. Not semaphore. Convert async function to sync version.

See [dispatch_group_notify](https://developer.apple.com/documentation/dispatch/1452933-dispatch_group_notify?language=objc)

**Comment by dreampiggy** [comment]
```objectivec
dispatch_queue_t targetQueue = dispatch_queue_create("com.targetQueue", DISPATCH_QUEUE_CONCURRENT);
dispatch_group_t group = dispatch_group_create();
dispatch_group_enter(group);
[[SDWebImageManager sharedManager] loadImageWithURL:[NSURL URLWithString:adModel.imageUrl] options:SDWebImageRefreshCached progress:nil completed:^(UIImage * _Nullable image, NSData * _Nullable data, NSError * _Nullable error, SDImageCacheType cacheType, BOOL finished, NSURL * _Nullable imageURL) {
	dispatch_group_leave(group);
}];
dispatch_group_notify(group, targetQueue, ^(void) {
	printf("finished");
});
```

**Comment by zhongwuzw** [comment]
@sunforcode I suggest you don't use lock to block any thread, it potentially dangerous. And `completed` block would not always be called, for example, you cancel all loading operations, so please don't do this. If you want to do some logical handler only if image be loaded, I suggest you do it in `completed` block, just like :

```
    [[SDWebImageManager sharedManager]loadImageWithURL:url options:SDWebImageRefreshCached progress:nil completed:^(UIImage * _Nullable image, NSData * _Nullable data, NSError * _Nullable error, SDImageCacheType cacheType, BOOL finished, NSURL * _Nullable imageURL) {
        __strong __typeof(wself) sself = wself;
        if (!sself) { return; }
        [sself handleCustomThingsIfLoadOperationCompleted:image:url:....];
    }];
```

**Comment by sunforcode** [comment]
> @sunforcode I suggest you don't use lock to block any thread, it potentially dangerous. And `completed` block would not always be called, for example, you cancel all loading operations, so please don't do this. If you want to do some logical handler only if image be loaded, I suggest you do it in `completed` block, just like :
> 
> ```
>     [[SDWebImageManager sharedManager]loadImageWithURL:url options:SDWebImageRefreshCached progress:nil completed:^(UIImage * _Nullable image, NSData * _Nullable data, NSError * _Nullable error, SDImageCacheType cacheType, BOOL finished, NSURL * _Nullable imageURL) {
>         __strong __typeof(wself) sself = wself;
>         if (!sself) { return; }
>         [sself handleCustomThingsIfLoadOperationCompleted:image:url:....];
>     }];
> ```

ty,i want to asyn request imagedata ,and block the main thread by use dispatch_semaphor ,and when request finished , the main thread continue,i use nsurlsession instead thank u anyway

**Comment by zhongwuzw** [comment]
@sunforcode Fine!
